src/openassetio-core/include/openassetio/specification/Specification.hpp

```diff
@@ -0,0 +1,127 @@
+ * integers, floating point, or booleans. Any of a trait's properties
+ * can be legitimately left unset - it is up to the consumer (host or
+ * manager, depending on the API method) to decide how this should be
+ * handled.
+ *
+ * @todo Add SimpleMap trait property value type.
+ *
+ * @see trait::property
+ *
+ * Various API methods require a populated specification to be provided
+ * by the host, which the manager can interrogate in order to determine
+ * the correct response.
+ *
+ * Conversely, various API methods, in particular @needsref
+ * ManagerInterface::resolve, require the manager to return a
+ * populated specification to the host. The traits (and hence their
+ * properties) contained within the returned specification are
+ * determined by the intersection of the traits that were requested by
+ * the host and the traits that the manager supports.
+ *
+ * Since specifications are generic dictionary-like data structures,
+ * accurate data access/mutation relies on well-known trait IDs and
+ * property names. This introduces a possible avenue for user error due
+ * to misspelling, as well as difficulty in discovering what properties
+ * may be available for a given trait.
+ *
+ * Therefore, it is strongly advised that accessing and mutating trait
+ * properties is performed using trait view wrapper classes wherever
+ * possible, rather than directly using the accessor/mutator functions
+ * on the specification.
+ *
+ * @see trait::TraitBase
+ *
+ */
+class OPENASSETIO_CORE_EXPORT Specification {
+ public:
+  /// List of supported trait IDs.
+  using TraitIds = std::vector<trait::TraitId>;
+
+  /**
+   * Construct such that this specification supports the given list of
+   * trait IDs.
+   *
+   * @param traitIds List of IDs of traits that this specification
+   * supports.
+   */
+  explicit Specification(const TraitIds& traitIds);
+
+  /**
+   * Defaulted virtual destructor establishing this as a base class
+   * for more concrete specifications.
+   */
+  virtual ~Specification();
+
+  /**
+   * Return whether this specification supports the given trait.
+   *
+   * @param traitId ID of trait to check for support.
+   * @return `true` if trait is supported, `false` otherwise.
+   */
+  [[nodiscard]] bool hasTrait(const trait::TraitId& traitId) const;
+
+  /**
+   * Get the value of a given trait property, if the property has
+   * been set.
+   *
+   * @param[out] out Storage fo result, only written to if the property
+   * is set.
+   * @param traitId ID of trait to query.
+   * @param propertyKey Key of trait's property to query.
+   * @return `true` if value was found, `false` if it is unset.
+   * @exception `std::out_of_range` if the trait is not supported by
+   * this specification.
+   */
+  [[nodiscard]] bool getTraitProperty(trait::property::Value* out, const trait::TraitId& traitId,
+                                      const trait::property::Key& propertyKey) const;
```

Comment:
[fix] We should probably also document what constitutes a valid trait name/property key

Reply:
I'm not sure what you mean, as in, it's any arbitrary string, as can be found by looking at the type.  Do we have further restrictions to add?  I guess the trait ID needs to be globally unique (I think I've called that out somewhere).

Reply:
Yeah, was wondering if its fine for them to contain any valid UTF-8 character, and how that then maps to method names, etc.

Reply:
Added some notes in c959d09f - I added a note to strongly recommend ASCII, which I'm not sure is really necessary these days, but would take some research to find out.

Reply:
GitHub won't let me resolve this comment 🤷‍♂️